---
title: "3chems DMP dmr and final correlation"
author: "Guannan Shen"
date: "March 25, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/ECCHO_github/DataProcessed/combp/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs, echo=FALSE }
## set up workspace
rm(list = ls())
library(data.table)
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(tibble)
library(openxlsx)
library(grid) # low-level grid functions are required
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# ls()
# slotNames(x)
# getSlots(x)

```

```{r dmp, echo=FALSE}
########### import dmp results ##################
f_pfhxs_DMP <- fread("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/genomewide_chem/2019-03-07_f_pfhxs__CpGs_withChem.csv", header = T)
f_pfoa_DMP <- fread("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/genomewide_chem/2019-03-07_f_pfoa__CpGs_withChem.csv", header = T)
f_pfos_DMP <- fread("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/genomewide_chem/2019-03-07_f_pfos__CpGs_withChem.csv", header = T)

m_pfhxs_DMP <- fread("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/genomewide_chem/2019-03-07_m_pfhxs__CpGs_withChem.csv", header = T)
m_pfoa_DMP <- fread("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/genomewide_chem/2019-03-07_m_pfoa__CpGs_withChem.csv", header = T)
m_pfos_DMP <- fread("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/genomewide_chem/2019-03-07_m_pfos__CpGs_withChem.csv", header = T)

### top 300

topdmp <- function(DMP, n) {
  topdmp = data.frame(DMP) %>% dplyr::select(-c("weights", "fdr", "sig")) %>%
                  dplyr::rename(rawp = raw) %>% dplyr::top_n(-n, rawp)
  name = substitute(DMP)
  write.xlsx(topdmp, paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/top300dmp/", sep = "_", name, n ,".xlsx" ) )
}

topdmp(f_pfhxs_DMP, 300)
topdmp(f_pfoa_DMP, 300)
topdmp(f_pfos_DMP, 300)

topdmp(m_pfhxs_DMP, 300)
topdmp(m_pfoa_DMP, 300)
topdmp(m_pfos_DMP, 300)



```


```{r combp}
########## DMR results from combp #################
f_pfhxs_dmr <- read.delim("f_pfhxs.anno.hg19.bed", header=TRUE, sep="\t")
f_pfoa_dmr <- read.delim("f_pfoa.anno.hg19.bed", header=TRUE, sep="\t")
f_pfos_dmr <- read.delim("f_pfos.anno.hg19.bed", header=TRUE, sep="\t")
m_pfhxs_dmr <- read.delim("m_pfhxs.anno.hg19.bed", header=TRUE, sep="\t")
m_pfoa_dmr <- read.delim("m_pfoa.anno.hg19.bed", header=TRUE, sep="\t")
m_pfos_dmr <- read.delim("m_pfos.anno.hg19.bed", header=TRUE, sep="\t")
head(f_pfhxs_dmr)

dmrformat <- function(dmr){
  ## this is only for combp results
  dmr_form = data.frame(dmr) %>% arrange(z_sidak_p) %>% 
                         dplyr::mutate(No. = 1:nrow(dmr) )  %>% 
                        dplyr::rename(chrom = X.chrom) %>%
                        select(-c("min_p", "z_p") ) %>%
                           select(No. , everything())
  name = substitute(dmr)
  write.xlsx(dmr_form, file = paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/combpdmr/", sep = "_", name,".xlsx" )
         )
}

dmrformat(f_pfhxs_dmr)
dmrformat(f_pfos_dmr)
dmrformat(f_pfoa_dmr)

dmrformat(m_pfhxs_dmr)
dmrformat(m_pfos_dmr)
dmrformat(m_pfoa_dmr)

```


```{r cpgsinDMR}
#get annotation;
anno = as.data.frame(getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19))

#get probes in DMR regions;

####### for combp, and DMRcate input 
########## also get the directionaliy of methylation by betafc (the beta fold change for given design)
  getprobes_perDMR = function(dmr, dmp, anno, pcutoff){
    name = paste(substitute(dmr), pcutoff, sep = "_")
    dmr_sig = data.frame(dmr) %>% arrange(z_sidak_p) %>% 
                         dplyr::mutate(No. = 1:nrow(dmr) )  %>% 
                        dplyr::rename(chrom = X.chrom) %>%
                        select(-c("min_p", "z_p") ) %>%
                           select(No. , everything()) %>% filter(z_sidak_p <= pcutoff)
    # chrom, start, end
    chrom = dmr_sig[, 2]; start =  dmr_sig[, 3]; stop =  dmr_sig[, 4]; n_dmr = nrow(dmr_sig)
    max_probes = max(dmr_sig$n_probes)
    
    # 2 means probename and rawp, + 3 means cpg1 and positive rate No_probes
    maxlength = (2*max_probes + 3)
    dmrprobes_all = matrix(NA, 0, maxlength)
    n_probes = NULL
  
    for (i in 1:n_dmr) {
      anno_probes = data.frame(anno) %>% dplyr::filter(chr == chrom[i]) %>%
                   dplyr::filter(pos >= start[i]) %>% dplyr::filter (pos <= stop[i]) %>% dplyr::select(Name)
      
      probes_dmp = data.frame(dmp) %>% 
                   filter(ID %in% anno_probes$Name)  %>% 
                      select(ID, betafc, raw ) %>% dplyr::rename(beta = betafc, raw_p = raw)
      probes = merge(anno_probes, probes_dmp, by.x = "Name", by.y = "ID")
      probes_order = data.frame(probes) %>% dplyr::arrange(raw_p )
      n_probes = c( n_probes, nrow(probes_order) )
      
    if (n_probes[i] == dmr_sig$n_probes[i] ) {
       dmrprobes = matrix(NA, 1, 0)
      for (j in 1: n_probes[i])
         {
          ## beta at column 2 
         dmrprobes = cbind(dmrprobes, probes_order[j, -2])
          colnames(dmrprobes)[ (2*j -1): (2*j )] = c(paste("Name", j, sep = "") ,
                                paste("raw_p", j, sep = "") )
         }
        dmrprobes = dmrprobes %>% dplyr::mutate(cpg1 = probes_order$Name[1],
                                     positive =  sum(probes_order$beta > 0)/n_probes[i],
                                     No_probes = n_probes[i] ) %>%
                             dplyr::select(cpg1, positive, No_probes, everything())
    # Convert a row of a data frame to vector
    dmrprobes = unlist(dmrprobes)
    # cbind or rbind different lengths vectors without repeating the elements of the shorter vectors
    length(dmrprobes) = maxlength
    }
      else {
        dmrprobes = c(probes_order$Name[1], 0  ,"Wrong Number of probes")
        names(dmrprobes) <- c("cpg1", "positive", "No_probes")
        # # cbind or rbind different lengths vectors without repeating the elements of the shorter vectors
        length(dmrprobes) = maxlength
      }
      
    dmrprobes_all = data.frame(rbind(dmrprobes_all, dmrprobes))
    row.names(dmrprobes_all) = NULL
    }
  fulllist = cbind(dmr_sig, dmrprobes_all)
  write.xlsx(fulllist, file =  paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/probesDMR/", sep = "_", name,"sigDMR.xlsx" ) )
  cpg1 = data.frame(fulllist) %>% dplyr::select(c("No.", "chrom", "start", "end",
                                                  "n_probes", "z_sidak_p", 
                                                  "cpg1", "positive", "No_probes"))
  
    write.xlsx(cpg1, file =  paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/", sep = "_", name,"topCpG.xlsx" ) )
    
}

getprobes_perDMR(f_pfhxs_dmr, f_pfhxs_DMP , anno, 0.05)
getprobes_perDMR(f_pfos_dmr, f_pfos_DMP , anno, 0.05)
getprobes_perDMR(f_pfoa_dmr, f_pfoa_DMP , anno, 0.05)

getprobes_perDMR(m_pfhxs_dmr, m_pfhxs_DMP , anno, 0.05)
getprobes_perDMR(m_pfos_dmr, m_pfos_DMP , anno, 0.05)
getprobes_perDMR(m_pfoa_dmr, m_pfoa_DMP , anno, 0.05)

```



```{r test, echo= F, eval= F}
#################### test #######################
pcutoff <- 0.05
    dmr_sig = data.frame(f_pfhxs_dmr) %>% arrange(z_sidak_p) %>% 
                         dplyr::mutate(No. = 1:nrow(f_pfhxs_dmr) )  %>% 
      
                        dplyr::rename(chrom = X.chrom) %>%
                        select(-c("min_p", "z_p") ) %>%
                           select(No. , everything()) %>% filter(z_sidak_p <= pcutoff)
    # chrom, start, end
    chrom = dmr_sig[, 2]; start =  dmr_sig[, 3]; stop =  dmr_sig[, 4]; n_dmr = nrow(dmr_sig)
    
    max_probes = max(dmr_sig$n_probes)
    
    # 2 means probename and rawp, + 3 means cpg1 and positive rate No_probes
    maxlength = (2*max_probes + 3)
    dmrprobes_all = matrix(NA, 0, maxlength)
    n_probes = NULL
  
    for (i in 1:n_dmr) {
      anno_probes = data.frame(anno) %>% dplyr::filter(chr == chrom[i]) %>%
                   dplyr::filter(pos >= start[i]) %>% dplyr::filter (pos <= stop[i]) %>% dplyr::select(Name)
      
      probes_dmp = data.frame(f_pfhxs_DMP) %>% 
        
                   filter(ID %in% anno_probes$Name)  %>% 
                      select(ID, betafc, raw ) %>% dplyr::rename(beta = betafc, raw_p = raw)
      probes = merge(anno_probes, probes_dmp, by.x = "Name", by.y = "ID")
      probes_order = data.frame(probes) %>% dplyr::arrange(raw_p )
      n_probes = c( n_probes, nrow(probes_order) )
  
  
    if (n_probes[i] == dmr_sig$n_probes[i] ) {
       dmrprobes = matrix(NA, 1, 0)
      for (j in 1: n_probes[i])
         {
          ## beta at column 2 
         dmrprobes = cbind(dmrprobes, probes_order[j, -2])
          colnames(dmrprobes)[ (2*j -1): (2*j )] = c(paste("Name", j, sep = "") ,
                                paste("raw_p", j, sep = "") )
         }
        dmrprobes = dmrprobes %>% dplyr::mutate(cpg1 = probes_order$Name[1],
                                     positive =  sum(probes_order$beta > 0)/n_probes[i],
                                     No_probes = n_probes[i] ) %>%
                             dplyr::select(cpg1, positive, No_probes, everything())
    # Convert a row of a data frame to vector
    dmrprobes = unlist(dmrprobes)
    # cbind or rbind different lengths vectors without repeating the elements of the shorter vectors
    length(dmrprobes) = maxlength
    }
      else {
        dmrprobes = c(probes_order$Name[1], 0  ,"Wrong Number of probes")
        names(dmrprobes) <- c("cpg1", "positive", "No_probes")
        # # cbind or rbind different lengths vectors without repeating the elements of the shorter vectors
        length(dmrprobes) = maxlength
      }
      
    dmrprobes_all = data.frame(rbind(dmrprobes_all, dmrprobes))
    row.names(dmrprobes_all) = NULL
    }
 fulllist = cbind(dmr, dmrprobes_all)

```

