---
title: "Scatter plots with ln(Chem)"
author: "Guannan Shen"
date: "March 27, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs, echo=FALSE }
## set up workspace
rm(list = ls())
library(data.table)
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(readxl)
library(mutoss)
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# ls()
# slotNames(x)
# getSlots(x)

```


```{r mval_and_conc, echo=FALSE}
####### import top cpg ################
top_f_pfhxs <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_f_pfhxs_dmr_0.05_topCpG.xlsx")
top_f_pfos <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_f_pfos_dmr_0.05_topCpG.xlsx")
top_f_pfoa <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_f_pfoa_dmr_0.05_topCpG.xlsx")

top_m_pfhxs <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_m_pfhxs_dmr_0.05_topCpG.xlsx")
top_m_pfos <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_m_pfos_dmr_0.05_topCpG.xlsx")
top_m_pfoa <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_m_pfoa_dmr_0.05_topCpG.xlsx")

showtop <- function(top){
  top1 = top  %>% dplyr::mutate(rate = round(as.numeric(positive), 4)) %>% 
            dplyr::select(-No_probes, -positive)  %>%
    dplyr::rename(Hypermethylation_prop. = rate)
  name = substitute(top)
  kable(top1, digits = c(1,1,1,1,1,15,1,4), caption = name)
}
showtop(top_f_pfhxs)
showtop(top_f_pfos)
showtop(top_f_pfoa)

showtop(top_m_pfhxs)
showtop(top_m_pfos)
showtop(top_m_pfoa)
########## get mval for topcpg
Mtop_f_pfhxs <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/Mval_top_f_pfhxs_.csv")
Mtop_f_pfos <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/Mval_top_f_pfos_.csv")
Mtop_f_pfoa <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/Mval_top_f_pfoa_.csv")

Mtop_m_pfhxs <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/Mval_top_m_pfhxs_.csv")
Mtop_m_pfos <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/Mval_top_m_pfos_.csv")
Mtop_m_pfoa <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/Mval_top_m_pfoa_.csv")

#### the same order as the pvalue of DMR, ascending
sum(colnames(Mtop_f_pfhxs)[-1] != top_f_pfhxs$cpg1)


########## get chem conc
# all clinical data, chemical conc, obesity outcomes
clinchem_f <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/clin_chem_f.csv")
clinchem_m <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/clin_chem_m.csv")
sum(Mtop_f_pfhxs$pid != clinchem_f$pid) != 0

```


```{r scatter}
doscatter <- function(mval, chem){
  if(sum(mval$pid != chem$pid) != 0)
  stop("Wrong pid order!")
  gender1 = unlist( strsplit(deparse(substitute(mval)), "_", fixed = T))  [2]
  gender2 = unlist( strsplit(deparse(substitute(chem)), "_", fixed = T))  [2]
  if(gender1 != gender2)
    {stop("Use same gender for both datasets")}
  else{
    gender = ifelse(gender1 == "f", "Female", "Male") 
  }
  chemname = unlist( strsplit(deparse(substitute(mval)), "_", fixed = T))  [3]
  xname = paste("ln", chemname, sep = "")
  x = chem[, colnames(chem) == xname]
  ncpg = ncol(mval) - 1
  for (i in 1:ncpg) {
    j = i+1
    cpg = colnames(mval)[j]
    ggplot( mval,aes(x, mval[, j]) ) +
          geom_point() +
            xlab(paste(xname,gender, sep = "_" ) ) +
            ylab(paste("M Values of ", sep = "", cpg) ) +
           geom_smooth(method = "lm", se = F) +
           theme_minimal()
    ggsave( paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/plots/scatter_chem/scatter", xname, cpg, gender,sep = "_", ".png" ) )
  }
}

doscatter(Mtop_f_pfhxs, clinchem_f)
doscatter(Mtop_f_pfos, clinchem_f)
doscatter(Mtop_f_pfoa, clinchem_f)

doscatter(Mtop_m_pfhxs, clinchem_m)
doscatter(Mtop_m_pfos, clinchem_m)
doscatter(Mtop_m_pfoa, clinchem_m)

```