---
title: "From DMR to Obesity Outcomes for 3 chemmicals"
author: "Guannan Shen"
date: "March 4, 2019"
output: 
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
  word_document:
    toc: yes
    toc_depth: '5'
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
# pid583_pfoa_pfos_pfhxs.csv
```

```{r libs}
## set up workspace
rm(list = ls())
library(data.table)
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(DMRcate)
library(parallel)
library(devtools)
# install_github("kvnkuang/pbmcapply", ref = "dev")
library(pbmcapply)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(tibble)
library(grid) # low-level grid functions are required
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# ls()
# slotNames(x)
# getSlots(x)

```

```{r dataprep}
#### all clinical data but chemical conc
pfas_cell_583 <- fread(file = "/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/pfas_cell_583.csv", 
                       header = T)
head(pfas_cell_583)
dim(pfas_cell_583)
#### chemical data and transform
chems <- fread(file = "/home/guanshim/Documents/gitlab/ECCHO_github/DataRaw/pid583_pfoa_pfos_pfhxs.csv", 
               header = T)

sum(chems$pid != pfas_cell_583$pid)
chems_ln <- chems %>% dplyr::mutate(lnpfhxs = log(PFHXS_ng_ml, base = exp(1)),
                                    lnpfoa = log(PFOA_ng_ml, base = exp(1)),
                                    lnpfos = log(PFOS_ng_ml, base = exp(1)))
head(chems_ln)
dim(chems_ln)
#### all data but mval
clin_chem <- inner_join(pfas_cell_583, chems_ln, by = "pid")
dim(clin_chem)
# by gender
clin_chem_f <- data.frame(clin_chem) %>% filter(infant_sex == "Female") 
dim(clin_chem_f)
# write.csv(clin_chem_f, 
#           "/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/clin_chem_f.csv",
#           row.names = F)
clin_chem_m <- data.frame(clin_chem) %>% filter(infant_sex == "Male") 
dim(clin_chem_m)
# write.csv(clin_chem_m, 
#           "/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/clin_chem_m.csv",
#           row.names = F)
## head(clin_chem)
# typeof(clin_chem)
# clin_chem <- data.table(clin_chem)

#### mval with each row represents a participant
t_mval <- fread( "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/t_mval.csv",
                 header = T)
head(t_mval[,1:5])
## merge two data.table
# t_mval$pid
# use data.table merge
# merge is much faster than inner_join
dt_all <- merge(clin_chem, t_mval, by = "pid")

dim(dt_all)
# # head(dt_all[,31:40])
# # data.table syntax
# dt_all <- data.table(dt_all)
# dt_all_f <-  dt_all[infant_sex == "Female"]
# # save female data
# dim(dt_all_f)
# fwrite(dt_all_f, "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/dt_all_f.csv", row.names = F)
dt_all_f <- fread( "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/dt_all_f.csv", header = T)
dim(dt_all_f)
# save male data 
dt_all_m <-  dt_all[infant_sex == "Male"]
# dim(dt_all_m)
# fwrite(dt_all_m, "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/dt_all_m.csv", row.names = F)
dt_all_m <- fread( "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/dt_all_m.csv", header = T)
dim(dt_all_m)

```


```{r genomewide fitting}
# if (requireNamespace("broom", quietly = TRUE)) {
#   iris %>%
#     group_by(Species) %>%
#     group_map(~ broom::tidy(lm(Petal.Length ~ Sepal.Length, data = .x)))
# }
############ since inner_join is slow, hold on trying the group_map #################

# formula + Race + Bcell + CD4T + CD8T + Gran + Mono + NK + nRBC

```