---
title: "Male Chem DMR Obesity"
author: "Guannan Shen"
date: "March 6, 2019"
output: 
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
  word_document:
    toc: yes
    toc_depth: '5'
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
# pid583_pfoa_pfos_pfhxs.csv
```

```{r libs}
## set up workspace
rm(list = ls())
library(data.table)
library(knitr)
library(tidyverse)
library(magrittr)
library(stats)
library(DMRcate)
library(parallel)
library(devtools)
install_github("kvnkuang/pbmcapply", ref = "dev")
library(pbmcapply)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(tibble)
library(grid) # low-level grid functions are required
options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# ls()
# slotNames(x)
# getSlots(x)

```


```{r dataprep}
## all female
# all clinical data, chemical conc, obesity outcomes
clin_chem_m <- read.csv("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/clin_chem_m.csv")
dim(clin_chem_m)
# mval
dt_all_m <- fread( "~/Documents/gitlab/ECCHO_github/DataProcessed/for_obesity/dt_all_m.csv", header = T)
dim(dt_all_m)
dt_all_m <- data.frame(dt_all_m)
# all pids were aligned 
sum(dt_all_m$pid != clin_chem_m$pid)
cpgs <- colnames(dt_all_m)[-c(1:26)]
# dt_all_m[,-c(1:26)]

```


## Pipeline to get the input for comb-p

```{r genomewide}
# local()  Ensure all your "strict" functions don't have globalenv as environment.
fchem <- local(function(mval, chem){
  fit1 = lm(mval ~ chem + Race + Bcell + CD4T + CD8T + Gran + Mono + NK + nRBC, 
                    data = clin_chem_m, na.action = na.omit) # Be sure to use the covariates for Anne's analysis
  fit.sum = summary(fit1)
  cpg.dat = fit.sum$coefficients[2, ] 
  return(cpg.dat)
})

#
# parallel
cl <- makeCluster(10)
clusterExport(cl, "clin_chem_m")
# clin_chem_m$lnpfoa clin_chem_m$lnpfhxs clin_chem_m$lnpfos
rslts.pfoa <- do.call("rbind", pbmclapply(dt_all_m[,-c(1:26)], FUN = fchem, clin_chem_m$lnpfoa, 
                                          mc.cores =getOption("mc.cores", 10L)))
# # 
# head(rslts.pfoa)
# # test 
# sum(rslts.cb.1[, 4] != rslts.pfoa[,4])
# colnames(rslts.pfoa)[4] <- "raw_p"
# sum(row.names(rslts.pfoa) != colnames(dt_all_m)[-c(1:26)])
# #
# rslts.pfoa <- data.frame(rslts.pfoa) %>% mutate(p_adjust = p.adjust(raw_p, method = "BH"),
#                                                 CpG = colnames(dt_all_m)[-c(1:26)])
rslts_format <- function(rslts.cb){
  rslts.cb = data.frame(rslts.cb)
  p_adjust = p.adjust(rslts.cb[, 4], method = "BH")
  rslts.cb = as.data.frame(cbind(rslts.cb, p_adjust))
  # rslts.cb = rslts.cb %>% arrange(`Pr(>|t|)`)
  colnames(rslts.cb)[4] = "rawp"
  cols.num = c(1:5)
  rslts.cb[cols.num] = sapply(rslts.cb[cols.num],as.character)
  rslts.cb[cols.num] = sapply(rslts.cb[cols.num],as.numeric)
}
################### work on Asus #########################
# can be used to find the top CpGs for scatter plot
pfoa_mval <- rslts_format(rslts.pfoa )
pfoa_mval <- data.frame(pfoa_mval)
pfoa_mval$CpG <- cpgs
# create the .bed file for comb-p
# the annotation for start position of .bed file
anno = as.data.frame(getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19))
bedfile <- function(chem_mval, chemname){
  # 
  paste("Might need to copy the .bed file to the home directory", sep = "\n",
        "chrom, chromStart, chromEnd, unadjusted, p_adjust, CpG")
  # Step 1. Merge rslts.cb with annotation
  chem_mval.merged = merge(chem_mval, anno, by.x = "CpG", by.y = "Name")
  chem_mval.merged$unadjusted = chem_mval.merged$rawp # Just the same column with a better name
  # Step 2. Generate BED style format
  chem_mval.bed = chem_mval.merged %>%
  mutate(stop = pos + 51) %>% # so generated one by adding 51 to position, 50 is the len of probe
  dplyr::rename(chrom = chr, chromStart = pos, chromEnd = stop) %>%
  select(chrom, chromStart, chromEnd, unadjusted, p_adjust, CpG)
  chem_mval.bed = data.frame(chem_mval.bed)
  write.table(chem_mval.bed, file= paste("~/Documents/gitlab/ECCHO_github/DataProcessed/genomewide_chem/", 
                                         chemname, "_male_mval.bed", sep = ""), 
              quote=F, sep="\t", row.names=F, col.names=F)
}

bedfile(chem_m = pfoa_mval, chemname = "lnpfoa")

## repeat the analysis for other chemical 
# clin_chem_m$lnpfoa clin_chem_m$lnpfhxs clin_chem_m$lnpfos
# genome wide association
rslts.pfhxs <- do.call("rbind", pbmclapply(dt_all_m[,-c(1:26)], FUN = fchem, clin_chem_m$lnpfhxs, 
                                          mc.cores =getOption("mc.cores", 10L)))
# get BH p-adjusted and reformat
pfhxs_mval <- rslts_format(rslts.pfhxs)
pfhxs_mval <- data.frame(fhxs_mval)
pfhxs_mval$CpG <- cpgs
# save .bed file
bedfile(chem_m = pfhxs_mval, chemname = "lnpfhxs")


############################# 
rslts.pfos <- do.call("rbind", pbmclapply(dt_all_m[,-c(1:26)], FUN = fchem, clin_chem_m$lnpfos, 
                                          mc.cores =getOption("mc.cores", 10L)))
# get BH p-adjusted and reformat
pfos_mval <- rslts_format(rslts.pfos)
pfos_mval <- data.frame(pfos_mval)
ppfos_mval$CpG <- cpgs
# save .bed file
bedfile(chem_m = pfos_mval, chemname = "lnpfos")

stopCluster(cl)
```


