---
title: "Pathways from 3 chemicals by gender Methylation"
author: "Guannan Shen"
date: "April 10, 2019"
output: 
  html_document:
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 5
    toc_float: yes
  word_document:
    toc: yes
    toc_depth: '5'
  pdf_document:
    latex_engine: lualatex
    number_sections: yes
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE, cache = FALSE}
require("knitr")
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
opts_chunk$set(engine = "R")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
## setting wd in DELL
## opts_knit$set(root.dir = "~/Stats/CIDA_OMICs/CIDA_OMICS/7659Stats_Genetics/HW5/")
## setting working directory in asus 
## opts_knit$set(root.dir = "C:/Users/hithr/Documents/Stats/gitlab/Cario_RNASeq_Microbiom_Inte/DataRaw/") 
## setting working directory in ubuntu
opts_knit$set(root.dir = "~/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/")
                                                 
## cache = F, if cache = T, will not revaluate code chunk everytime
## double or more space to insert a line break
```


```{r libs, echo=FALSE }
## set up workspace
rm(list = ls())

library(knitr)
library(tidyverse)
library(enrichR)
library(openxlsx)
library(stats)
library(readxl)

options(stringsAsFactors = F)
options(dplyr.width = Inf)
getwd()
## not in function
'%nin%' <- Negate('%in%')

# ######## clean memory ######################
# rm(list = ls())
# gc()
# ls()
# slotNames(x)
# getSlots(x)

```

## Import Data


```{r data, echo=FALSE }
# load top1 cpg associated gene from FDR 0.05 DMR
####### import top cpg ################
top_f_pfhxs <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_f_pfhxs_dmr_0.05_topCpG.xlsx")
top_f_pfos <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_f_pfos_dmr_0.05_topCpG.xlsx")
top_f_pfoa <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_f_pfoa_dmr_0.05_topCpG.xlsx")

top_m_pfhxs <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_m_pfhxs_dmr_0.05_topCpG.xlsx")
top_m_pfos <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_m_pfos_dmr_0.05_topCpG.xlsx")
top_m_pfoa <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3chems_results_data/topProbeDMR/_m_pfoa_dmr_0.05_topCpG.xlsx")
########### import top 300 cpgs from DMP ##########
f_pfhxs_DMP_300 <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/ECCHO_3Chems_fe_male_pathway/top300dmp/_f_pfhxs_DMP_300_.xlsx")
f_pfos_DMP_300 <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/ECCHO_3Chems_fe_male_pathway/top300dmp/_f_pfos_DMP_300_.xlsx")
f_pfoa_DMP_300 <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/ECCHO_3Chems_fe_male_pathway/top300dmp/_f_pfoa_DMP_300_.xlsx")

m_pfhxs_DMP_300 <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/ECCHO_3Chems_fe_male_pathway/top300dmp/_m_pfhxs_DMP_300_.xlsx")
m_pfos_DMP_300 <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/ECCHO_3Chems_fe_male_pathway/top300dmp/_m_pfos_DMP_300_.xlsx")
m_pfoa_DMP_300 <- read_excel("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/ECCHO_3Chems_fe_male_pathway/top300dmp/_m_pfoa_DMP_300_.xlsx")

######################### import dataset of top500 dmp and top1 CpG of p<1 dmr ######################
folder3 <- "/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3_chems_top500_dmr1/dmr_1/"
folder4 <- "/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/3_chems_top500_dmr1/top500dmp/"
######### dmr cutoff at 1 from combp #########
top_f_pfhxs_1 <- read_excel(paste(folder3, "_1_f_pfhxs_dmr_1_topCpG.xlsx", sep = ""))
top_f_pfos_1 <- read_excel(paste(folder3, "_1_f_pfos_dmr_1_topCpG.xlsx", sep = ""))
top_f_pfoa_1 <- read_excel(paste(folder3, "_1_f_pfoa_dmr_1_topCpG.xlsx", sep = ""))

top_m_pfhxs_1 <- read_excel(paste(folder3, "_1_m_pfhxs_dmr_1_topCpG.xlsx", sep = ""))
top_m_pfos_1 <- read_excel(paste(folder3, "_1_m_pfos_dmr_1_topCpG.xlsx", sep = ""))
top_m_pfoa_1 <- read_excel(paste(folder3, "_1_m_pfoa_dmr_1_topCpG.xlsx", sep = ""))
########## top 500 dmp 
f_pfhxs_DMP_500 <- read_excel(paste(folder4, "_f_pfhxs_DMP_500_.xlsx", sep = ""))
f_pfos_DMP_500 <- read_excel(paste(folder4, "_f_pfos_DMP_500_.xlsx", sep = ""))
f_pfoa_DMP_500 <- read_excel(paste(folder4, "_f_pfoa_DMP_500_.xlsx", sep = ""))

m_pfhxs_DMP_500 <- read_excel(paste(folder4, "_m_pfhxs_DMP_500_.xlsx", sep = ""))
m_pfos_DMP_500 <- read_excel(paste(folder4, "_m_pfos_DMP_500_.xlsx", sep = ""))
m_pfoa_DMP_500 <- read_excel(paste(folder4, "_m_pfoa_DMP_500_.xlsx", sep = ""))
## Enrich R pathway analysis  

```

## Funciton to make ErichR Query 


```{r function, echo=F}
# getpathway_top <- function(CpGtable, database, gene_col){
#   #########3 CpGtable a string and database a vector ############## 
#   ######## gene_col a string
#   name = CpGtable
#   # run enrichr
#   pathways = enrichr( eval( call("$" ,as.name(CpGtable), gene_col) ), database)
#   # data
#   go_mol = data.frame( pathways[["GO_Molecular_Function_2018"]] ) 
#   go_bio = data.frame( pathways[["GO_Biological_Process_2018"]] )
#   go_cel = data.frame( pathways[["GO_Cellular_Component_2018"]] )
#   kegg = data.frame( pathways[["KEGG_2019_Human"]] )
#   
#   ## save 
#   write.xlsx(kegg, 
#              paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/pathways_eccho_3chem/", sep = "_", name, "kegg_2019.xlsx" ) )
#   write.xlsx(go_bio, 
#              paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/pathways_eccho_3chem/", sep = "_", name, "go_biopro_2018.xlsx" ) )
#   write.xlsx(go_mol, 
#              paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/pathways_eccho_3chem/", sep = "_", name, "go_mole_2018.xlsx" ) )
#   write.xlsx(go_cel, 
#              paste("/home/guanshim/Documents/gitlab/ECCHO_github/DataProcessed/pathways_eccho_3chem/", sep = "_", name, "go_cell_2018.xlsx" ) )
#   
#   #return
#   return( pathways )
# }
```


```{r format}
getpathway_top <- function(CpGtable, database, gene_col){
  #########3 CpGtable a string and database a vector ############## 
  ######## gene_col a string
  name = CpGtable
  # run enrichr
  pathways = enrichr( eval( call("$" ,as.name(CpGtable), gene_col) ), database)
  # data
  kegg = cbind (data.frame( pathways[["KEGG_2019_Human"]] ), c("KEGG_2019_Human")) 
  colnames(kegg)[10] = "Database"
  lap_kegg = matrix(unlist(strsplit(kegg$Overlap, "/", fixed = TRUE)), ncol = 2, byrow = T)[ , 1]
  kegg_3 = kegg[ lap_kegg >= 2, ]
  
  go_mol = cbind (data.frame( pathways[["GO_Molecular_Function_2018"]] ) , c("GO_Molecular_Function_2018"))
  colnames(go_mol)[10] = "Database"
  lap_go_mol = matrix(unlist(strsplit(go_mol$Overlap, "/", fixed = TRUE)), ncol = 2, byrow = T)[ , 1]
  go_mol_3 = go_mol[ lap_go_mol >= 2, ]
  
  go_bio = cbind ( data.frame( pathways[["GO_Biological_Process_2018"]] ) , c("GO_Biological_Process_2018"))
  colnames(go_bio)[10] = "Database"
  lap_go_bio = matrix(unlist(strsplit(go_bio$Overlap, "/", fixed = TRUE)), ncol = 2, byrow = T)[ , 1]
  go_bio_3 = go_bio[ lap_go_bio >= 2, ]
  
  go_cel = cbind ( data.frame( pathways[["GO_Cellular_Component_2018"]] ) , c("GO_Cellular_Component_2018"))
  colnames(go_cel)[10] = "Database"
  lap_go_cel = matrix(unlist(strsplit(go_cel$Overlap, "/", fixed = TRUE)), ncol = 2, byrow = T)[ , 1]
  go_cel_3 = go_cel[ lap_go_cel >= 2, ]
  
  data = rbind(kegg_3, go_mol_3, go_bio_3, go_cel_3) %>% dplyr::select( -c("Old.P.value", 
                                                                           "Old.Adjusted.P.value") ) %>%
                                                    dplyr::filter(Adjusted.P.value <= 0.2)
  ## save 
  
  #return
  return(data )
}
```


```{r testformat, include=F, eval=F}
dbs <- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018", "GO_Biological_Process_2018",
         "KEGG_2019_Human")
top1_gene <- c("refGene_name")
top300_gene <- c("UCSC_RefGene_Name")

top1_data <- c("top_f_pfhxs", "top_f_pfos","top_f_pfoa", 
               "top_m_pfhxs", "top_m_pfos","top_m_pfoa")
top300_data <- c("f_pfhxs_DMP_300", "f_pfos_DMP_300", "f_pfoa_DMP_300",
                 "m_pfhxs_DMP_300", "m_pfos_DMP_300", "m_pfoa_DMP_300")


getpathway_top( "top_f_pfhxs",  dbs, top1_gene) %>% kable(caption = "female pfhxs Top1 DMR" )



```

## Results

### Two Filters

1. At least 2 genes are in the pathway.  

2. The adjusted p value (BH, FDR) is less than or equal to 0.2.  


```{r run, include=F, eval=F}
dbs <- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018", "GO_Biological_Process_2018",
         "KEGG_2019_Human")
top1_gene <- c("refGene_name")
top300_gene <- c("UCSC_RefGene_Name")

top1_data <- c("top_f_pfhxs", "top_f_pfos","top_f_pfoa", 
               "top_m_pfhxs", "top_m_pfos","top_m_pfoa")
top300_data <- c("f_pfhxs_DMP_300", "f_pfos_DMP_300", "f_pfoa_DMP_300",
                 "m_pfhxs_DMP_300", "m_pfos_DMP_300", "m_pfoa_DMP_300")


############# run #################
f_pfhxs_1 <- getpathway_top( "top_f_pfhxs",  dbs, top1_gene)
kable(f_pfhxs_1$KEGG_2019_Human[1:5, ], 
      caption = "female pfhxs Top1 DMR: KEGG_2019_Human" )
kable(f_pfhxs_1$GO_Biological_Process_2018[1:5, ], 
      caption = "female pfhxs Top1 DMR: GO_Biological_Process_2018" )
kable(f_pfhxs_1$GO_Molecular_Function_2018[1:5, ], 
      caption = "female pfhxs Top1 DMR: GO_Molecular_Function_2018" )
kable(f_pfhxs_1$GO_Cellular_Component_2018[1:5, ], 
      caption = "female pfhxs Top1 DMR: GO_Cellular_Component_2018" )

f_pfhxs_300 <- getpathway_top( "f_pfhxs_DMP_300",  dbs, top300_gene)
kable(f_pfhxs_300$KEGG_2019_Human[1:5, ], 
      caption = "female pfhxs Top300 DMP: KEGG_2019_Human" )
kable(f_pfhxs_300$GO_Biological_Process_2018[1:5, ], 
      caption = "female pfhxs Top300 DMP: GO_Biological_Process_2018" )
kable(f_pfhxs_300$GO_Molecular_Function_2018[1:5, ], 
      caption = "female pfhxs Top300 DMP: GO_Molecular_Function_2018" )
kable(f_pfhxs_300$GO_Cellular_Component_2018[1:5, ], 
      caption = "female pfhxs Top300 DMP: GO_Cellular_Component_2018" )

f_pfos_1 <- getpathway_top( "top_f_pfos",  dbs, top1_gene)
kable(f_pfos_1$KEGG_2019_Human[1:5, ], 
      caption = "female pfos Top1 DMR: KEGG_2019_Human" )
kable(f_pfos_1$GO_Biological_Process_2018[1:20, ], 
      caption = "female pfos Top1 DMR: GO_Biological_Process_2018" )
kable(f_pfos_1$GO_Molecular_Function_2018[1:10, ], 
      caption = "female pfos Top1 DMR: GO_Molecular_Function_2018" )
kable(f_pfos_1$GO_Cellular_Component_2018[1:5, ], 
      caption = "female pfos Top1 DMR: GO_Cellular_Component_2018" )

f_pfos_300 <- getpathway_top( "f_pfos_DMP_300",  dbs, top300_gene)
kable(f_pfos_300$KEGG_2019_Human[1:5, ], 
      caption = "female pfos Top300 DMP: KEGG_2019_Human" )
kable(f_pfos_300$GO_Biological_Process_2018[1:5, ], 
      caption = "female pfos Top300 DMP: GO_Biological_Process_2018" )
kable(f_pfos_300$GO_Molecular_Function_2018[1:5, ], 
      caption = "female pfos Top300 DMP: GO_Molecular_Function_2018" )
kable(f_pfos_300$GO_Cellular_Component_2018[1:5, ], 
      caption = "female pfos Top300 DMP: GO_Cellular_Component_2018" )

f_pfoa_1 <- getpathway_top( "top_f_pfoa",  dbs, top1_gene)
kable(f_pfoa_1$KEGG_2019_Human[1:5, ], 
      caption = "female pfoa Top1 DMR: KEGG_2019_Human" )
kable(f_pfoa_1$GO_Biological_Process_2018[1:5, ], 
      caption = "female pfoa Top1 DMR: GO_Biological_Process_2018" )
kable(f_pfoa_1$GO_Molecular_Function_2018[1:5, ], 
      caption = "female pfoa Top1 DMR: GO_Molecular_Function_2018" )
kable(f_pfoa_1$GO_Cellular_Component_2018[1:5, ], 
      caption = "female pfoa Top1 DMR: GO_Cellular_Component_2018" )

f_pfoa_300 <- getpathway_top( "f_pfoa_DMP_300",  dbs, top300_gene)
kable(f_pfoa_300$KEGG_2019_Human[1:5, ], 
      caption = "female pfoa Top300 DMP: KEGG_2019_Human" )
kable(f_pfoa_300$GO_Biological_Process_2018[1:5, ], 
      caption = "female pfoa Top300 DMP: GO_Biological_Process_2018" )
kable(f_pfoa_300$GO_Molecular_Function_2018[1:5, ], 
      caption = "female pfoa Top300 DMP: GO_Molecular_Function_2018" )
kable(f_pfoa_300$GO_Cellular_Component_2018[1:5, ], 
      caption = "female pfoa Top300 DMP: GO_Cellular_Component_2018" )

```


```{r male, include= F, eval=F}
############# run #################
m_pfhxs_1 <- getpathway_top( "top_m_pfhxs",  dbs, top1_gene)
kable(m_pfhxs_1$KEGG_2019_Human[1:5, ], 
      caption = "male pfhxs Top1 DMR: KEGG_2019_Human" )
kable(m_pfhxs_1$GO_Biological_Process_2018[1:5, ], 
      caption = "male pfhxs Top1 DMR: GO_Biological_Process_2018" )
kable(m_pfhxs_1$GO_Molecular_Function_2018[1:5, ], 
      caption = "male pfhxs Top1 DMR: GO_Molecular_Function_2018" )
kable(m_pfhxs_1$GO_Cellular_Component_2018[1:5, ], 
      caption = "male pfhxs Top1 DMR: GO_Cellular_Component_2018" )

m_pfhxs_300 <- getpathway_top( "m_pfhxs_DMP_300",  dbs, top300_gene)
kable(m_pfhxs_300$KEGG_2019_Human[1:5, ], 
      caption = "male pfhxs Top300 DMP: KEGG_2019_Human" )
kable(m_pfhxs_300$GO_Biological_Process_2018[1:5, ], 
      caption = "male pfhxs Top300 DMP: GO_Biological_Process_2018" )
kable(m_pfhxs_300$GO_Molecular_Function_2018[1:5, ], 
      caption = "male pfhxs Top300 DMP: GO_Molecular_Function_2018" )
kable(m_pfhxs_300$GO_Cellular_Component_2018[1:5, ], 
      caption = "male pfhxs Top300 DMP: GO_Cellular_Component_2018" )

m_pfos_1 <- getpathway_top( "top_m_pfos",  dbs, top1_gene)
kable(m_pfos_1$KEGG_2019_Human[1:5, ], 
      caption = "male pfos Top1 DMR: KEGG_2019_Human" )
kable(m_pfos_1$GO_Biological_Process_2018[1:20, ], 
      caption = "male pfos Top1 DMR: GO_Biological_Process_2018" )
kable(m_pfos_1$GO_Molecular_Function_2018[1:10, ], 
      caption = "male pfos Top1 DMR: GO_Molecular_Function_2018" )
kable(m_pfos_1$GO_Cellular_Component_2018[1:5, ], 
      caption = "male pfos Top1 DMR: GO_Cellular_Component_2018" )

m_pfos_300 <- getpathway_top( "m_pfos_DMP_300",  dbs, top300_gene)
kable(m_pfos_300$KEGG_2019_Human[1:5, ], 
      caption = "male pfos Top300 DMP: KEGG_2019_Human" )
kable(m_pfos_300$GO_Biological_Process_2018[1:5, ], 
      caption = "male pfos Top300 DMP: GO_Biological_Process_2018" )
kable(m_pfos_300$GO_Molecular_Function_2018[1:5, ], 
      caption = "male pfos Top300 DMP: GO_Molecular_Function_2018" )
kable(m_pfos_300$GO_Cellular_Component_2018[1:5, ], 
      caption = "male pfos Top300 DMP: GO_Cellular_Component_2018" )

m_pfoa_1 <- getpathway_top( "top_m_pfoa",  dbs, top1_gene)
kable(m_pfoa_1$KEGG_2019_Human[1:5, ], 
      caption = "male pfoa Top1 DMR: KEGG_2019_Human" )
kable(m_pfoa_1$GO_Biological_Process_2018[1:5, ], 
      caption = "male pfoa Top1 DMR: GO_Biological_Process_2018" )
kable(m_pfoa_1$GO_Molecular_Function_2018[1:5, ], 
      caption = "male pfoa Top1 DMR: GO_Molecular_Function_2018" )
kable(m_pfoa_1$GO_Cellular_Component_2018[1:5, ], 
      caption = "male pfoa Top1 DMR: GO_Cellular_Component_2018" )

m_pfoa_300 <- getpathway_top( "m_pfoa_DMP_300",  dbs, top300_gene)
kable(m_pfoa_300$KEGG_2019_Human[1:20, ], 
      caption = "male pfoa Top300 DMP: KEGG_2019_Human" )
kable(m_pfoa_300$GO_Biological_Process_2018[1:5, ], 
      caption = "male pfoa Top300 DMP: GO_Biological_Process_2018" )
kable(m_pfoa_300$GO_Molecular_Function_2018[1:5, ], 
      caption = "male pfoa Top300 DMP: GO_Molecular_Function_2018" )
kable(m_pfoa_300$GO_Cellular_Component_2018[1:20, ], 
      caption = "male pfoa Top300 DMP: GO_Cellular_Component_2018" )

```


### Female Enrichment Results from Top500 DMP or TOP1 CpG from p less than 1 DMR

```{r female500}
dbs <- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018", "GO_Biological_Process_2018",
         "KEGG_2019_Human")
top1_gene <- c("refGene_name")
top500_gene <- c("UCSC_RefGene_Name")


getpathway_top( "top_f_pfhxs_1",  dbs, top1_gene) %>% kable(caption = "female pfhxs Top1 from less than 1 DMR" )

getpathway_top( "f_pfhxs_DMP_500",  dbs, top500_gene) %>% kable(caption = "female pfhxs Top500 DMP" )

getpathway_top( "top_f_pfos_1",  dbs, top1_gene) %>% kable(caption = "female pfos Top1 from less than 1 DMR" )

getpathway_top( "f_pfos_DMP_500",  dbs, top500_gene) %>% kable(caption = "female pfos Top500 DMP" )

getpathway_top( "top_f_pfoa_1",  dbs, top1_gene) %>% kable(caption = "female pfoa Top1 from less than 1 DMR" )

getpathway_top( "f_pfoa_DMP_500",  dbs, top500_gene) %>% kable(caption = "female pfoa Top500 DMP" )
 




```

### Male Enrichment Results from Top500 DMP or TOP1 CpG from p less than 1 DMR

```{r male500}
dbs <- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018", "GO_Biological_Process_2018",
         "KEGG_2019_Human")
top1_gene <- c("refGene_name")
top500_gene <- c("UCSC_RefGene_Name")

getpathway_top( "top_m_pfhxs_1",  dbs, top1_gene) %>% kable(caption = "male pfhxs Top1 from less than 1 DMR" )

getpathway_top( "m_pfhxs_DMP_500",  dbs, top500_gene) %>% kable(caption = "male pfhxs Top500 DMP" )

getpathway_top( "top_m_pfos_1",  dbs, top1_gene) %>% kable(caption = "male pfos Top1 from less than 1 DMR" )

getpathway_top( "m_pfos_DMP_500",  dbs, top500_gene) %>% kable(caption = "male pfos Top500 DMP" )

getpathway_top( "top_m_pfoa_1",  dbs, top1_gene) %>% kable(caption = "male pfoa Top1 from less than 1 DMR" )

getpathway_top( "m_pfoa_DMP_500",  dbs, top500_gene) %>% kable(caption = "male pfoa Top500 DMP" )
```


